File: python/topological_sort.py
"""
Topological sorting for Directed Acyclic Graphs (DAGs).

Produces a linear ordering of vertices such that for every directed edge (u, v),
vertex u comes before v in the ordering. Uses both DFS-based and Kahn's algorithm
(BFS-based) approaches for different use cases.

Time complexity: O(V + E) for both algorithms, where V is vertices and E is edges.
Space complexity: O(V + E) for the graph representation and auxiliary data structures.
"""

from __future__ import annotations

from collections import deque

# Don't use annotations during contest
from typing import Generic, TypeVar

NodeT = TypeVar("NodeT")


class TopologicalSort(Generic[NodeT]):
    def __init__(self) -> None:
        self.graph: dict[NodeT, list[NodeT]] = {}
        self.in_degree: dict[NodeT, int] = {}

    def add_edge(self, u: NodeT, v: NodeT) -> None:
        """Add directed edge from u to v."""
        if u not in self.graph:
            self.graph[u] = []
            self.in_degree[u] = 0
        if v not in self.in_degree:
            self.in_degree[v] = 0
            self.graph[v] = []

        self.graph[u].append(v)
        self.in_degree[v] += 1

    def kahn_sort(self) -> list[NodeT] | None:
        """
        Topological sort using Kahn's algorithm (BFS-based).

        Returns the topological ordering, or None if the graph has a cycle.
        """
        in_deg = self.in_degree.copy()
        queue = deque([node for node, deg in in_deg.items() if deg == 0])
        result = []

        while queue:
            node = queue.popleft()
            result.append(node)

            for neighbor in self.graph[node]:
                in_deg[neighbor] -= 1
                if in_deg[neighbor] == 0:
                    queue.append(neighbor)

        # Check if all nodes are processed (no cycle)
        if len(result) != len(self.in_degree):
            return None

        return result

    def dfs_sort(self) -> list[NodeT] | None:
        """
        Topological sort using DFS.

        Returns the topological ordering, or None if the graph has a cycle.
        """
        WHITE, GRAY, BLACK = 0, 1, 2
        color = dict.fromkeys(self.in_degree, WHITE)
        result = []

        def dfs(node: NodeT) -> bool:
            if color[node] == GRAY:  # Back edge (cycle)
                return False
            if color[node] == BLACK:  # Already processed
                return True

            color[node] = GRAY
            for neighbor in self.graph[node]:
                if not dfs(neighbor):
                    return False

            color[node] = BLACK
            result.append(node)
            return True

        for node in self.in_degree:
            if color[node] == WHITE and not dfs(node):
                return None

        return result[::-1]

    def has_cycle(self) -> bool:
        """Check if the graph contains a cycle."""
        return self.kahn_sort() is None

    def longest_path(self) -> dict[NodeT, int]:
        """
        Find longest path from each node in the DAG.

        Returns a dictionary mapping each node to its longest path length.
        """
        topo_order = self.kahn_sort()
        if topo_order is None:
            msg = "Graph contains a cycle"
            raise ValueError(msg)

        dist = dict.fromkeys(self.in_degree, 0)

        for node in topo_order:
            for neighbor in self.graph[node]:
                dist[neighbor] = max(dist[neighbor], dist[node] + 1)

        return dist


def test_main() -> None:
    ts: TopologicalSort[int] = TopologicalSort()
    edges = [(5, 2), (5, 0), (4, 0), (4, 1), (2, 3), (3, 1)]
    for u, v in edges:
        ts.add_edge(u, v)

    kahn_result = ts.kahn_sort()
    dfs_result = ts.dfs_sort()

    assert kahn_result is not None
    assert dfs_result is not None
    assert not ts.has_cycle()

    # Test with cycle
    ts_cycle: TopologicalSort[int] = TopologicalSort()
    ts_cycle.add_edge(1, 2)
    ts_cycle.add_edge(2, 3)
    ts_cycle.add_edge(3, 1)
    assert ts_cycle.has_cycle()


# Don't write tests below during competition.


def test_empty_graph() -> None:
    ts: TopologicalSort[int] = TopologicalSort()

    # Empty graph should return empty list
    assert ts.kahn_sort() == []
    assert ts.dfs_sort() == []
    assert not ts.has_cycle()


def test_single_node() -> None:
    ts: TopologicalSort[str] = TopologicalSort()
    ts.add_edge("A", "A")  # This creates a self-loop (cycle)

    assert ts.has_cycle()
    assert ts.kahn_sort() is None
    assert ts.dfs_sort() is None

    # Single node without self-loop
    ts2: TopologicalSort[str] = TopologicalSort()
    ts2.in_degree["A"] = 0
    ts2.graph["A"] = []

    result = ts2.kahn_sort()
    assert result == ["A"]
    assert not ts2.has_cycle()


def test_simple_chain() -> None:
    # Linear chain: A -> B -> C -> D
    ts: TopologicalSort[str] = TopologicalSort()
    edges = [("A", "B"), ("B", "C"), ("C", "D")]
    for u, v in edges:
        ts.add_edge(u, v)

    kahn_result = ts.kahn_sort()
    dfs_result = ts.dfs_sort()

    assert kahn_result == ["A", "B", "C", "D"]
    assert dfs_result == ["A", "B", "C", "D"]
    assert not ts.has_cycle()


def test_simple_cycle() -> None:
    # Simple 3-cycle: A -> B -> C -> A
    ts: TopologicalSort[str] = TopologicalSort()
    edges = [("A", "B"), ("B", "C"), ("C", "A")]
    for u, v in edges:
        ts.add_edge(u, v)

    assert ts.has_cycle()
    assert ts.kahn_sort() is None
    assert ts.dfs_sort() is None


def test_disconnected_components() -> None:
    # Two disconnected chains: A->B and C->D
    ts: TopologicalSort[str] = TopologicalSort()
    edges = [("A", "B"), ("C", "D")]
    for u, v in edges:
        ts.add_edge(u, v)

    kahn_result = ts.kahn_sort()
    dfs_result = ts.dfs_sort()

    # Both should work, order might vary for disconnected components
    assert kahn_result is not None
    assert dfs_result is not None
    assert len(kahn_result) == 4
    assert len(dfs_result) == 4
    assert not ts.has_cycle()

    # Check that ordering constraints are satisfied
    assert kahn_result.index("A") < kahn_result.index("B")
    assert kahn_result.index("C") < kahn_result.index("D")


def test_diamond_dag() -> None:
    # Diamond: A -> B,C  B,C -> D
    ts: TopologicalSort[str] = TopologicalSort()
    edges = [("A", "B"), ("A", "C"), ("B", "D"), ("C", "D")]
    for u, v in edges:
        ts.add_edge(u, v)

    kahn_result = ts.kahn_sort()
    dfs_result = ts.dfs_sort()

    assert kahn_result is not None
    assert dfs_result is not None
    assert not ts.has_cycle()

    # Check ordering constraints
    for result in [kahn_result, dfs_result]:
        assert result.index("A") < result.index("B")
        assert result.index("A") < result.index("C")
        assert result.index("B") < result.index("D")
        assert result.index("C") < result.index("D")


def test_complex_dag() -> None:
    # More complex DAG with multiple valid orderings
    ts: TopologicalSort[int] = TopologicalSort()
    edges = [(1, 2), (1, 3), (2, 4), (3, 4), (3, 5), (4, 6), (5, 6)]
    for u, v in edges:
        ts.add_edge(u, v)

    kahn_result = ts.kahn_sort()
    dfs_result = ts.dfs_sort()

    assert kahn_result is not None
    assert dfs_result is not None
    assert not ts.has_cycle()

    # Verify all ordering constraints
    constraints = [(1, 2), (1, 3), (2, 4), (3, 4), (3, 5), (4, 6), (5, 6)]
    for result in [kahn_result, dfs_result]:
        for u, v in constraints:
            assert result.index(u) < result.index(v)


def test_large_cycle() -> None:
    # Large cycle: 0 -> 1 -> 2 -> ... -> 99 -> 0
    ts: TopologicalSort[int] = TopologicalSort()
    n = 100
    for i in range(n):
        ts.add_edge(i, (i + 1) % n)

    assert ts.has_cycle()
    assert ts.kahn_sort() is None
    assert ts.dfs_sort() is None


def test_tree_structure() -> None:
    # Binary tree structure (DAG)
    ts: TopologicalSort[int] = TopologicalSort()
    edges = [(1, 2), (1, 3), (2, 4), (2, 5), (3, 6), (3, 7)]
    for u, v in edges:
        ts.add_edge(u, v)

    kahn_result = ts.kahn_sort()
    dfs_result = ts.dfs_sort()

    assert kahn_result is not None
    assert dfs_result is not None
    assert not ts.has_cycle()

    # Check tree constraints
    for result in [kahn_result, dfs_result]:
        assert result.index(1) < result.index(2)
        assert result.index(1) < result.index(3)
        assert result.index(2) < result.index(4)
        assert result.index(2) < result.index(5)
        assert result.index(3) < result.index(6)
        assert result.index(3) < result.index(7)


def test_longest_path() -> None:
    # Test longest path functionality
    ts: TopologicalSort[str] = TopologicalSort()
    edges = [("A", "B"), ("A", "C"), ("B", "D"), ("C", "D"), ("D", "E")]
    for u, v in edges:
        ts.add_edge(u, v)

    longest_paths = ts.longest_path()

    # Expected longest paths from each node
    assert longest_paths["A"] == 0
    assert longest_paths["B"] == 1
    assert longest_paths["C"] == 1
    assert longest_paths["D"] == 2
    assert longest_paths["E"] == 3


def test_longest_path_with_cycle() -> None:
    # Should raise error for graphs with cycles
    ts: TopologicalSort[int] = TopologicalSort()
    edges = [(1, 2), (2, 3), (3, 1)]  # Cycle
    for u, v in edges:
        ts.add_edge(u, v)

    try:
        ts.longest_path()
        assert False, "Should raise ValueError for cyclic graph"
    except ValueError:
        pass


def test_multiple_sources() -> None:
    # Graph with multiple sources (nodes with in-degree 0)
    ts: TopologicalSort[int] = TopologicalSort()
    edges = [(1, 3), (2, 3), (3, 4), (5, 6)]
    for u, v in edges:
        ts.add_edge(u, v)

    kahn_result = ts.kahn_sort()
    dfs_result = ts.dfs_sort()

    assert kahn_result is not None
    assert dfs_result is not None
    assert not ts.has_cycle()

    # Sources (1, 2, 5) should come before their dependents
    for result in [kahn_result, dfs_result]:
        assert result.index(1) < result.index(3)
        assert result.index(2) < result.index(3)
        assert result.index(3) < result.index(4)
        assert result.index(5) < result.index(6)


def test_course_prerequisites() -> None:
    # Real-world example: course prerequisites
    ts: TopologicalSort[str] = TopologicalSort()
    # Math1 -> Math2 -> Math3, Physics1 -> Physics2, Math2 -> Physics2
    edges = [
        ("Math1", "Math2"), ("Math2", "Math3"),
        ("Physics1", "Physics2"), ("Math2", "Physics2")
    ]
    for u, v in edges:
        ts.add_edge(u, v)

    kahn_result = ts.kahn_sort()
    assert kahn_result is not None
    assert not ts.has_cycle()

    # Verify prerequisites
    assert kahn_result.index("Math1") < kahn_result.index("Math2")
    assert kahn_result.index("Math2") < kahn_result.index("Math3")
    assert kahn_result.index("Physics1") < kahn_result.index("Physics2")
    assert kahn_result.index("Math2") < kahn_result.index("Physics2")


def test_complex_dependencies() -> None:
    # Complex dependency graph
    ts: TopologicalSort[str] = TopologicalSort()
    edges = [
        ("underwear", "pants"), ("underwear", "shoes"),
        ("pants", "belt"), ("pants", "shoes"),
        ("shirt", "belt"), ("shirt", "tie"),
        ("tie", "jacket"), ("belt", "jacket"),
        ("socks", "shoes")
    ]
    for u, v in edges:
        ts.add_edge(u, v)

    kahn_result = ts.kahn_sort()
    dfs_result = ts.dfs_sort()

    assert kahn_result is not None
    assert dfs_result is not None
    assert not ts.has_cycle()

    # Check some key dependencies
    for result in [kahn_result, dfs_result]:
        assert result.index("underwear") < result.index("pants")
        assert result.index("pants") < result.index("shoes")
        assert result.index("socks") < result.index("shoes")
        assert result.index("shirt") < result.index("tie")
        assert result.index("tie") < result.index("jacket")


def test_stress_large_dag() -> None:
    # Large DAG: layered graph
    ts: TopologicalSort[int] = TopologicalSort()
    layers = 10
    nodes_per_layer = 10

    # Connect each node in layer i to all nodes in layer i+1
    for layer in range(layers - 1):
        for i in range(nodes_per_layer):
            for j in range(nodes_per_layer):
                u = layer * nodes_per_layer + i
                v = (layer + 1) * nodes_per_layer + j
                ts.add_edge(u, v)

    kahn_result = ts.kahn_sort()
    dfs_result = ts.dfs_sort()

    assert kahn_result is not None
    assert dfs_result is not None
    assert not ts.has_cycle()
    assert len(kahn_result) == layers * nodes_per_layer


def main() -> None:
    test_main()
    test_empty_graph()
    test_single_node()
    test_simple_chain()
    test_simple_cycle()
    test_disconnected_components()
    test_diamond_dag()
    test_complex_dag()
    test_large_cycle()
    test_tree_structure()
    test_longest_path()
    test_longest_path_with_cycle()
    test_multiple_sources()
    test_course_prerequisites()
    test_complex_dependencies()
    test_stress_large_dag()


if __name__ == "__main__":
    main()

File: cpp/topological_sort.cpp
/*
Topological sorting for Directed Acyclic Graphs (DAGs).

Produces a linear ordering of vertices such that for every directed edge (u, v),
vertex u comes before v in the ordering. Uses both DFS-based and Kahn's algorithm
(BFS-based) approaches for different use cases.

Time complexity: O(V + E) for both algorithms, where V is vertices and E is edges.
Space complexity: O(V + E) for the graph representation and auxiliary data structures.
*/

#include <iostream>
#include <vector>
#include <map>
#include <queue>
#include <algorithm>
#include <optional>
#include <stdexcept>
#include <cassert>

template<typename NodeT>
class TopologicalSort {
private:
    std::map<NodeT, std::vector<NodeT>> graph;
    std::map<NodeT, int> in_degree;

    enum Color { WHITE, GRAY, BLACK };

    bool dfs_helper(NodeT node, std::map<NodeT, Color>& color, std::vector<NodeT>& result) {
        if (color[node] == GRAY) {  // Back edge (cycle)
            return false;
        }
        if (color[node] == BLACK) {  // Already processed
            return true;
        }

        color[node] = GRAY;
        for (const auto& neighbor : graph[node]) {
            if (!dfs_helper(neighbor, color, result)) {
                return false;
            }
        }

        color[node] = BLACK;
        result.push_back(node);
        return true;
    }

public:
    void add_edge(NodeT u, NodeT v) {
        if (graph.find(u) == graph.end()) {
            graph[u] = {};
            in_degree[u] = 0;
        }
        if (in_degree.find(v) == in_degree.end()) {
            in_degree[v] = 0;
            graph[v] = {};
        }

        graph[u].push_back(v);
        in_degree[v]++;
    }

    std::optional<std::vector<NodeT>> kahn_sort() {
        /*
        Topological sort using Kahn's algorithm (BFS-based).

        Returns the topological ordering, or nullopt if the graph has a cycle.
        */
        std::map<NodeT, int> in_deg = in_degree;
        std::queue<NodeT> q;

        for (const auto& [node, deg] : in_deg) {
            if (deg == 0) {
                q.push(node);
            }
        }

        std::vector<NodeT> result;

        while (!q.empty()) {
            NodeT node = q.front();
            q.pop();
            result.push_back(node);

            for (const auto& neighbor : graph[node]) {
                in_deg[neighbor]--;
                if (in_deg[neighbor] == 0) {
                    q.push(neighbor);
                }
            }
        }

        // Check if all nodes are processed (no cycle)
        if (result.size() != in_degree.size()) {
            return std::nullopt;
        }

        return result;
    }

    std::optional<std::vector<NodeT>> dfs_sort() {
        /*
        Topological sort using DFS.

        Returns the topological ordering, or nullopt if the graph has a cycle.
        */
        std::map<NodeT, Color> color;
        for (const auto& [node, _] : in_degree) {
            color[node] = WHITE;
        }

        std::vector<NodeT> result;

        for (const auto& [node, _] : in_degree) {
            if (color[node] == WHITE && !dfs_helper(node, color, result)) {
                return std::nullopt;
            }
        }

        std::reverse(result.begin(), result.end());
        return result;
    }

    bool has_cycle() {
        return !kahn_sort().has_value();
    }

    std::map<NodeT, int> longest_path() {
        /*
        Find longest path from each node in the DAG.

        Returns a map from each node to its longest path length.
        */
        auto topo_order = kahn_sort();
        if (!topo_order.has_value()) {
            throw std::runtime_error("Graph contains a cycle");
        }

        std::map<NodeT, int> dist;
        for (const auto& [node, _] : in_degree) {
            dist[node] = 0;
        }

        for (const auto& node : topo_order.value()) {
            for (const auto& neighbor : graph[node]) {
                dist[neighbor] = std::max(dist[neighbor], dist[node] + 1);
            }
        }

        return dist;
    }
};

void test_main() {
    TopologicalSort<int> ts;
    std::vector<std::pair<int, int>> edges = {{5, 2}, {5, 0}, {4, 0}, {4, 1}, {2, 3}, {3, 1}};
    for (const auto& [u, v] : edges) {
        ts.add_edge(u, v);
    }

    auto kahn_result = ts.kahn_sort();
    auto dfs_result = ts.dfs_sort();

    assert(kahn_result.has_value());
    assert(dfs_result.has_value());
    assert(!ts.has_cycle());

    // Test with cycle
    TopologicalSort<int> ts_cycle;
    ts_cycle.add_edge(1, 2);
    ts_cycle.add_edge(2, 3);
    ts_cycle.add_edge(3, 1);
    assert(ts_cycle.has_cycle());
}

// Don't write tests below during competition.

void test_empty_graph() {
    TopologicalSort<int> ts;

    // Empty graph should return empty list
    assert(ts.kahn_sort().value().empty());
    assert(ts.dfs_sort().value().empty());
    assert(!ts.has_cycle());
}

void test_single_node_self_loop() {
    TopologicalSort<std::string> ts;
    ts.add_edge("A", "A");  // This creates a self-loop (cycle)

    assert(ts.has_cycle());
    assert(!ts.kahn_sort().has_value());
    assert(!ts.dfs_sort().has_value());
}

void test_single_node_no_edges() {
    TopologicalSort<int> ts;
    ts.add_edge(1, 2);  // Add a simple edge to create node

    TopologicalSort<int> ts2;
    // Can't easily test single isolated node, skip this case
}

void test_linear_chain() {
    TopologicalSort<int> ts;
    ts.add_edge(1, 2);
    ts.add_edge(2, 3);
    ts.add_edge(3, 4);
    ts.add_edge(4, 5);

    auto kahn_result = ts.kahn_sort();
    auto dfs_result = ts.dfs_sort();

    assert(kahn_result.has_value());
    assert(dfs_result.has_value());
    assert(!ts.has_cycle());

    // Check ordering
    auto kahn = kahn_result.value();
    assert(kahn.size() == 5);
    // Verify 1 comes before 2, 2 before 3, etc.
    for (size_t i = 0; i < kahn.size() - 1; i++) {
        assert(kahn[i] < kahn[i + 1]);
    }
}

void test_multiple_sources() {
    TopologicalSort<int> ts;
    ts.add_edge(1, 3);
    ts.add_edge(2, 3);
    ts.add_edge(3, 4);

    auto result = ts.kahn_sort();
    assert(result.has_value());
    assert(!ts.has_cycle());

    // Both 1 and 2 should come before 3
    auto vec = result.value();
    auto pos_1 = std::find(vec.begin(), vec.end(), 1) - vec.begin();
    auto pos_2 = std::find(vec.begin(), vec.end(), 2) - vec.begin();
    auto pos_3 = std::find(vec.begin(), vec.end(), 3) - vec.begin();
    auto pos_4 = std::find(vec.begin(), vec.end(), 4) - vec.begin();

    assert(pos_1 < pos_3);
    assert(pos_2 < pos_3);
    assert(pos_3 < pos_4);
}

void test_diamond_shape() {
    TopologicalSort<std::string> ts;
    ts.add_edge("A", "B");
    ts.add_edge("A", "C");
    ts.add_edge("B", "D");
    ts.add_edge("C", "D");

    auto kahn = ts.kahn_sort();
    auto dfs = ts.dfs_sort();

    assert(kahn.has_value());
    assert(dfs.has_value());
    assert(!ts.has_cycle());

    // A should come first, D should come last
    auto kahn_vec = kahn.value();
    assert(kahn_vec.front() == "A");
    assert(kahn_vec.back() == "D");
}

void test_complex_cycle() {
    TopologicalSort<int> ts;
    ts.add_edge(1, 2);
    ts.add_edge(2, 3);
    ts.add_edge(3, 4);
    ts.add_edge(4, 2);  // Creates cycle 2->3->4->2

    assert(ts.has_cycle());
    assert(!ts.kahn_sort().has_value());
    assert(!ts.dfs_sort().has_value());
}

void test_disconnected_components() {
    TopologicalSort<int> ts;
    // Component 1
    ts.add_edge(1, 2);
    ts.add_edge(2, 3);
    // Component 2
    ts.add_edge(4, 5);
    ts.add_edge(5, 6);

    auto result = ts.kahn_sort();
    assert(result.has_value());
    assert(!ts.has_cycle());
    assert(result.value().size() == 6);
}

void test_longest_path() {
    TopologicalSort<int> ts;
    ts.add_edge(1, 2);
    ts.add_edge(1, 3);
    ts.add_edge(2, 4);
    ts.add_edge(3, 4);
    ts.add_edge(4, 5);

    auto dist = ts.longest_path();

    assert(dist[1] == 0);
    assert(dist[5] == 3);  // Longest path: 1->2->4->5 or 1->3->4->5
}

void test_longest_path_with_cycle() {
    TopologicalSort<int> ts;
    ts.add_edge(1, 2);
    ts.add_edge(2, 3);
    ts.add_edge(3, 1);

    bool caught = false;
    try {
        ts.longest_path();
    } catch (const std::runtime_error&) {
        caught = true;
    }
    assert(caught);
}

void test_comparison_kahn_vs_dfs() {
    TopologicalSort<int> ts;
    ts.add_edge(5, 2);
    ts.add_edge(5, 0);
    ts.add_edge(4, 0);
    ts.add_edge(4, 1);
    ts.add_edge(2, 3);
    ts.add_edge(3, 1);

    auto kahn = ts.kahn_sort();
    auto dfs = ts.dfs_sort();

    assert(kahn.has_value());
    assert(dfs.has_value());

    // Both should produce valid topological orderings
    // (may differ, but both should be valid)
}

void test_large_graph() {
    TopologicalSort<int> ts;

    // Create a chain of 1000 nodes
    for (int i = 0; i < 999; i++) {
        ts.add_edge(i, i + 1);
    }

    auto result = ts.kahn_sort();
    assert(result.has_value());
    assert(result.value().size() == 1000);
    assert(!ts.has_cycle());
}

void test_string_nodes() {
    TopologicalSort<std::string> ts;
    ts.add_edge("undershirt", "shirt");
    ts.add_edge("pants", "belt");
    ts.add_edge("shirt", "belt");
    ts.add_edge("belt", "jacket");
    ts.add_edge("socks", "shoes");
    ts.add_edge("pants", "shoes");

    auto result = ts.kahn_sort();
    assert(result.has_value());
    assert(!ts.has_cycle());
}

int main() {
    test_empty_graph();
    test_single_node_self_loop();
    test_linear_chain();
    test_multiple_sources();
    test_diamond_shape();
    test_complex_cycle();
    test_disconnected_components();
    test_longest_path();
    test_longest_path_with_cycle();
    test_comparison_kahn_vs_dfs();
    test_large_graph();
    test_string_nodes();
    test_main();
    std::cout << "All Topological Sort tests passed!" << std::endl;
    return 0;
}
